Implement support for caddy placeholders in replacement

fanste opened the pull request.

fanste committed:
Implement support for caddy placeholders in replacement

htfy96 commented on filter.go:
```diff
@@ -22,6 +23,8 @@ func init() {
 	httpcaddyfile.RegisterHandlerDirective("filter", parseCaddyfile)
 }
 
+var paramReplacementPattern = regexp.MustCompile("\\{[a-zA-Z0-9_\\-.]+}")
```
Can you make the `+` non-greedy by changing it to `+?`? I would expect it to be faster if we use non-greedy matches

fanste replied:
According to some online blog posts, that should not change anything, as the group only allows a limited set of characters. But I will change it, if you feel more comfortable with it.

htfy96 commented on filter.go:
```diff
@@ -173,10 +212,13 @@ func (m Middleware) ServeHTTP(w http.ResponseWriter, r *http.Request, next caddy
 			return fmt.Errorf("failed to read from response body: %w", err)
 		}
-		replaced := m.compiledSearchRegex.ReplaceAll(buf.Bytes(), []byte(m.Replacement))
+		replaced := m.compiledSearchRegex.ReplaceAllFunc(buf.Bytes(), func(input []byte) []byte {
```
Perhaps I missed something, but is it possible to generate the replacement string ahead of time?

For example:

```go
// When m.Replacement is "abc{system.os}def"
realReplacement := paramReplacementPattern.ReplaceAllFunc(replacement, func(input2 []byte) []byte {
		return m.paramReplacer(input2, groups, repl)
	})
// realReplacement is "abclinuxdef"
replaced := m.compiledSearchRegex.ReplaceAll(buf.Bytes(), realReplacement)
```

fanste replied:
That makes no sense, as the caddy placeholders also contain request-specific, reverse-proxy-specific, ... placeholders. If we generate the replacement string ahead of time (I think you mean inside `Provision`?), than we won't be able to use those placeholders. And thats the usecase that I need. The code is btw similar to the one of caddy v1 filter module.

htfy96 replied:
I see. Thanks for explaining this

htfy96 commented on filter.go:
```diff
@@ -160,6 +163,42 @@ func (csr *CappedSizeRecorder) WriteHeader(statusCode int) {
+	return replacement
+}
+
+func (m Middleware) paramReplacer(input []byte, groups [][]byte, repl *caddy.Replacer) []byte {
```
Add some document here saying it supports both `{caddy_replacer}` and capture group replacement like `{1}`

fanste replied:
Done, also added a comment to the README

htfy96 reviewed (changes requested):
Thanks for contribution. Can you add some test cases in `test/` directory covering both `{caddy_replacement}` and `{1}` use cases?

fanste committed:
Apply improvements

 - non-greedy replacement-pattern
 - store 'replacement' as byte-array instead of string
 - improve documentation

fanste committed:
Add test for new replacement handling

fanste commented:
> 
> 
> Thanks for contribution. Can you add some test cases in `test/` directory covering both `{caddy_replacement}` and `{1}` use cases?

I have no linux machine or emulator to test the `run-test` script, but I tried to verify the test implementation using windows tools

htfy96 reviewed (approved):
lgtm. Thanks for spending time improving this!
